### ISSUE TITLE
Writing time history variables to Paraview

### ISSUE THREAD

**Issue by oriolcg** [gridap/Gridap.jl#268]
In what follows I outline a tentative code to output time-history variables to Paraview.
@santiagobadia, what do you think about this (see the main issue commented in the code)?

```julia
using WriteVTK

struct PVDObject # Not sure how to name it
    filebase::String
    pdvFile::CollectionFile
    # Constructor
    PVDObject(filebase::String) = new(filebase, paraview_collection(filebase))
end

function start(filebase::String)
    pvdObject = PVDObject(filebase)
end

function _get_current_celldata(step::Int, celldata::Dict)
    current_celldata = Dict()
    for key,value in celldata
        current_celldata[key] = value[step]
    end
    current_celldata
end

function _get_current_cellfields(step::Int, cellfields::Dict)
    current_cellfields = Dict()
    for key,value in cellfields
        current_cellfields[key] = value[step]
    end
    current_cellfields
end

# Add a set of time steps to the pvd
# This function should be extended for all the cases defined for Gridap's function `writevtk`. The code below is only a particular implementation
# I use arrays for all entries to handle problems with evolving meshes (adaptivity, Lagrangian, ...), but we could have a function with a fixed triangulation
function add!(pvdObject::PVDObject, times::Array{Real,1}, trians::Array{Triangulation,1}; orders=[-1], nsubcellsVector=[-1], celldata=Dict(), cellfields=Dict())

    for (step,time) in enumerate(times)

        trian = trians[step]
        order = orders[step]
        nsubcells = nsubcellsVector[step]
        current_celldata = _get_current_celldata(step,celldata)
        current_cellfields = _get_current_cellfields(step,cellfields)

        # ISSUE 1:
        # I would like to reuse as much as possible all the functions in Vtk.jl, but
        # is not possible keeping the current implementation of all the `writevtk` functions.
        # Now `writevtk` outputs a string array with the names of the files. Here we
        # want the actual vtkFile generated within the function `write_vtk_file`, which
        # is not available outside this function.
        vtkFile = writevtk_or_similar(trian, filebase, pvdObject.filebase, order, nsubcells, current_celldata, current_cellfields)
        pvdObject.pvdFile[time] = vtkFile
    end
end

function end(pvdObject::PVDObject)
    vtk_save(pvdObject.pvdFile)
end
```

The user would do something like:
```julia
xh = sol_t[1]
times = sol_t[2]
pvd = start("whateverfilename")
add!(pvd, times, trian, cellfields=["solution"=>xh])
end(pvd)
```

**Comment by fverdugo** [comment]
Hi @oriolcg ! thanks for the draft.

I would try to implement a more "lazy" approach that does not require to build an array or triangulations, one per time step etc.

I am wondering if we need to implement anything at all, or perhaps a small fix in our `writevtk` would be enough.

Let's assume that we fix our `writevtk` so-that it returns the handle `vtkfile` returned by the internal call to `vtk_grid`.  That is: adding `return vtkfile` just before this line:
https://github.com/gridap/Gridap.jl/blob/1fde6966ca7cf5fdb62597a309e2a5f6f1c3701c/src/Visualization/Vtk.jl#L110



Then, we are ready to go with the machinery provided by `WriteVTK`. The API would be like this:

```julia
using Gridap
using WriteVTK

...

pvd = paraview_collection("my_pvd_file")

steps = solve(op)
for (i,(uh, t)) in enumerate(steps)
  vtkfile = writevtk(trian,"results_$i",cellfields=["uh"=>uh])
 pvd[t] = vtkfile
end

vtk_save(pvd)
```

or using do-block syntax:


```julia
using Gridap
using WriteVTK

...

steps = solve(op)
paraview_collection("my_pvd_file") do pvd

  for (i,(uh, t)) in enumerate(steps)
    vtkfile = writevtk(trian,"results_$i",cellfields=["uh"=>uh])
   pvd[t] = vtkfile
  end

end

```

**Comment by fverdugo** [comment]
I often prefer to use do-block syntax when handling a resource that needs to be initialized and finalized (the pvd object in this case). It is more safe, since the object will be finalized even if there is an error in the function that fills the object.

**Comment by fverdugo** [comment]
Perhaps, we need a function called `createvtk` that works like `writevtk` but it just returns the `vtkfile` handler without writing anything to disk, since this will be perhaps done by `paraview_collection`. @oriolcg can you double check this?

**Comment by oriolcg** [comment]
Thanks for the comments @fverdugo! The approach you propose seems very reasonable and easy to implement. We just need to decide which is the best way to solve the issue with the `vtkfile`.

> Perhaps, we need a function called `createvtk` that works like `writevtk` but it just returns the `vtkfile` handler without writing anything to disk, since this will be perhaps done by `paraview_collection`. @oriolcg can you double check this?

Yes, the `vtk_save` is called within the `collection_add_timestep` function, i.e. when doing ```pvd[time] = vtkfile```. As you said, we need an equivalent function to `write_vtk_file` where the  line https://github.com/gridap/Gridap.jl/blob/1fde6966ca7cf5fdb62597a309e2a5f6f1c3701c/src/Visualization/Vtk.jl#L109 is replaced by something like ```return  vtkfile```. 

One way of doing that would be to define a `create_vtk_file` function and the corresponding `createvtk` functions, equivalent to all `writevtk` implementations. But this would lead to a lot of code replication and I don't think it's optimal. Do you have any idea on how to properly do that?

**Comment by fverdugo** [comment]
@oriolcg code replication is easy to solve in this case. Everything should be implemented in new functions `create_vtk_file` and `createvtk` (almost 1:1 of what we have now), and then implement the `write_vtk_file` and `writevtk` as a call to  `create_vtk_file` or `createvtk`  plus `vtk_save`.

could you try to do this?

**Comment by fverdugo** [comment]
Closed via PR #273
